libflux/core/src/semantic/convert.rs

```diff
@@ -242,42 +242,311 @@ fn convert_monotype(
+                    _ => {},
+                }
+            }
+            for param in func.opt {
+                match param.1 {
+                    MonoType::Var(e) =>  {
+                        let tmpvar = tvars_mapped.get(&e);
+                        match tmpvar {
+                            Some(vr) => {
+                                let res = recs.get(&vr);
+                                match res {
+                                    Some(rs) => {opt.insert(param.0, rs.clone());},
+                                    _ => {},
+                                }
+                            },
+                            _ => {},
+                        }
+                    }
+                    _ => {},
+                }
+            }
+            match func.pipe {
+                Some(prop) => {
+                    pipe = Some(types::Property {
+                        k: prop.k,
+                        v: convert_polytype_record_constraints(prop.v, tvars_mapped.clone(), recs.clone()),
+                    })
+                },
+                None => {},
+            }
+            MonoType::Fun(Box::new(types::Function {
+                req,
+                opt,
+                pipe: pipe,
+                retn: convert_polytype_record_constraints(func.retn, tvars_mapped.clone(), recs.clone()),
+            }))
+        }
+        MonoType::Record(rec) => {
+            let mut r = MonoType::Record(Box::new(types::Record::Empty));
+            match *rec {
+                types::Record::Empty => {}, 
+                types::Record::Extension{head: h, tail: t} => {
+                    let new_property = types::Property {
+                        k: h.k, 
+                        v: convert_polytype_record_constraints(h.v, tvars_mapped.clone(), recs.clone())
+                    };
+                    r = MonoType::Record(Box::new(types::Record::Extension {
+                        head: new_property,
+                        tail: convert_polytype_record_constraints(t, tvars_mapped.clone(), recs.clone()),
+                    }))
+                }
+            }
+            r
+        }
+    }
+}
+
 pub fn convert_polytype(
     type_expression: ast::TypeExpression,
     f: &mut Fresher,
 ) -> Result<types::PolyType> {
     let mut tvars = HashMap::<String, types::Tvar>::new();
-    let expr = convert_monotype(type_expression.monotype, &mut tvars, f)?;
+    let mut tvars_mapped = HashMap::<types::Tvar, types::Tvar>::new();
+    let mut expr = convert_monotype(type_expression.monotype, &mut tvars, f)?;
     let mut vars = Vec::<types::Tvar>::new();
-    let mut cons = SemanticMap::<types::Tvar, Vec<types::Kind>>::new();
-
-    for (name, tvar) in tvars {
-        vars.push(tvar);
-        let mut kinds = Vec::<types::Kind>::new();
+    let mut recs = HashMap::<types::Tvar, MonoType>::new();
+    
+    let row_tvar = f.fresh();
+    vars.push(row_tvar);
+    for (name, tvar) in &tvars {
```

Comment:
Rather than iterating over the `tvars` map, I might instead iterate over the constraints of the type expression, build a record constraint for each type variable, and then add a new mapping to `recs`.